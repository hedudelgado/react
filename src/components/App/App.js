import React, { Component } from 'react';
import {
  array,
  object,
  func,
  bool,
  node,
} from 'prop-types';
import {
  PASSWORD_VIEW,
  PAYMENT_CARD_VIEW,
  USER_DETAILS_VIEW,
  ROOM_REQUIREMENTS_VIEW,
  EXTRAS_PREFERENCES_VIEW,
} from './appActions';
import { ComponentTogglerContainer } from '../ComponentToggler';
import UserProfile from '../UserProfile';
import Password from '../Password';
import UserProfileFormContainer from '../UserProfileForm';
import PasswordFormContainer from '../PasswordForm';
import { PaymentCardContainer } from '../PaymentCard';
import { RegularGuestsContainer } from '../RegularGuests';
import { NewsletterContainer } from '../Newsletter';
import Modal from '../../layout/Modal';
import PaymentCardFormContainer from '../PaymentCardForm';
import { DeleteCustomerContainer } from '../DeleteCustomer';
import RoomRequirements from '../RoomRequirements';
import RoomRequirementsFormContainer from '../RoomRequirementsForm';
import ExtrasPreferences from '../ExtrasPreferences';
import ExtrasPreferencesFormContainer from '../ExtrasPreferencesForm';
import BaseStyle from '../../layout/BaseStyle';
import MemorableWord from '../MemorableWord';
import MoveAccount from '../MoveAccount';
import {
  STAYER, SELF, managePermissionFlag, USER_LOGGED_IN_EVENT,
} from '../../config.json';
import ManagePermission from '../ManagePermission';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLoggedIn = this.handleLoggedIn.bind(this);
  }

  componentDidMount() {
    window.addEventListener('message', this.handleLoggedIn);
  }

  componentWillUnmount() {
    window.removeEventListener('message', this.handleLoggedIn);
  }

  async handleLoggedIn(event) {
    if (event.data.action === USER_LOGGED_IN_EVENT) {
      const { getSettings } = this.props;
      const { contactDetail: { email }, sessionId, business } = JSON.parse(event.data.profile);
      getSettings(email, sessionId, !!business);
    }
  }

  render() {
    const {
      countries,
      dictionary,
      profile,
      regions,
      newsletter,
      displayModal,
      handleModalClose,
      modalContent,
      userDataReceived,
      application: {
        showEmailPreferences,
      },
    } = this.props;

    const {
      business,
      business: {
        accessLevel = '',
        tetheredGuid = '',
      } = {},
    } = profile;

    const appReady = userDataReceived
      && (!showEmailPreferences || !!regions.length)
      && !!countries.length
      && (!showEmailPreferences || !!Object.keys(newsletter).length)
      && !!Object.keys(dictionary).length
      && !!Object.keys(profile).length;

    const showExtrasPreferences = !(accessLevel === SELF || accessLevel === STAYER);

    return (
      <div data-test="AccountSettings">
        {appReady
        && (
          <BaseStyle className="wb-typography app-container">
            <ComponentTogglerContainer
              viewName={USER_DETAILS_VIEW}
              view={UserProfile}
              edit={UserProfileFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PASSWORD_VIEW}
              view={Password}
              edit={PasswordFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PAYMENT_CARD_VIEW}
              view={PaymentCardContainer}
              edit={PaymentCardFormContainer}
            />
            {managePermissionFlag && (
              <ManagePermission dictionary={dictionary} />
            )}
            <RegularGuestsContainer />
            {showEmailPreferences && (
              <NewsletterContainer />
            )}
            <ComponentTogglerContainer
              viewName={ROOM_REQUIREMENTS_VIEW}
              view={RoomRequirements}
              edit={RoomRequirementsFormContainer}
            />
            {showExtrasPreferences && (
              <ComponentTogglerContainer
                viewName={EXTRAS_PREFERENCES_VIEW}
                view={ExtrasPreferences}
                edit={ExtrasPreferencesFormContainer}
              />
            )}
            {!!tetheredGuid && (
              <MemorableWord />
            )}
            {business && !tetheredGuid && (
              <MoveAccount />
            )}
            {!business && (
            <DeleteCustomerContainer />
            )}
            {displayModal && <Modal data-test="Modal" display content={modalContent} handleClose={handleModalClose} />}
          </BaseStyle>
        )
      }
      </div>
    );
  }
}

App.propTypes = {
  countries: array.isRequired,
  dictionary: object.isRequired,
  profile: object.isRequired,
  getSettings: func.isRequired,
  displayModal: bool.isRequired,
  handleModalClose: func.isRequired,
  modalContent: node,
  newsletter: object.isRequired,
  regions: array.isRequired,
  userDataReceived: bool.isRequired,
  application: object.isRequired,
};

export default App;
