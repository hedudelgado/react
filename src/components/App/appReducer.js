import {
  SET_NOTIFICATION_MESSAGE, SET_EDIT_VIEW, DEFAULT_VIEW,
} from './appActions';

const defaultState = {
  editViewName: DEFAULT_VIEW,
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case SET_EDIT_VIEW:
      return {
        ...state,
        editViewName: action.editViewName,
      };
    case SET_NOTIFICATION_MESSAGE:
      return {
        ...state,
        notificationMessage: action.notificationMessage,
      };
    default:
      return state;
  }
};
