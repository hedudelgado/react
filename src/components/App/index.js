export { default } from './App';

export { default as AppActions } from './appActions';

export { default as AppContainer } from './AppContainer';

export { default as modalReducer } from './modalReducer';
