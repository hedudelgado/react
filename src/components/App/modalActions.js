export const MODAL_OPEN = 'MODAL_OPEN';
export const MODAL_CLOSE = 'MODAL_CLOSE';

export const openModal = content => ({
  type: MODAL_OPEN,
  content,
});

export const closeModal = () => ({
  type: MODAL_CLOSE,
});
