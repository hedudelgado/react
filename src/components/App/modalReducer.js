import {
  MODAL_OPEN,
  MODAL_CLOSE,
} from './modalActions';

export default (state = {
  display: false,
  content: null,
}, action) => {
  const {
    type,
    content,
  } = action;

  switch (type) {
    case MODAL_OPEN:
      return {
        ...state,
        display: true,
        content,
      };
    case MODAL_CLOSE:
      return {
        ...state,
        display: false,
        content: null,
      };
    default:
      return state;
  }
};
