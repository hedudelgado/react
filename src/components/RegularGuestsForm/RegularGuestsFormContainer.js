import React, { Component, Fragment, createRef } from 'react';
import {
  object,
  array,
  func,
  string,
  number,
} from 'prop-types';
import { Formik } from 'formik';
import Typography from '../../layout/Typography';
import BaseStyle from '../../layout/BaseStyle';
import Notification from '../../layout/Notification';
import { matchCountry } from '../../utils/matchCountry';
import { matchDialingCode } from '../../utils/matchDialingCode';
import RegularGuestsForm from './RegularGuestsForm';
import {
  REGULAR_GUESTS_VIEW,
  DEFAULT_VIEW,
  REGULAR_GUESTS_ERROR,
} from '../App/appActions';
import {
  validateTitle,
  validateFirstName,
  validateLastName,
  validateEmail,
  validatePassportNumber,
  validatePhoneNumber,

} from '../../utils/validations';
import { defaultCountry } from '../../config.json';

class RegularGuestsFormContainer extends Component {
  constructor(props) {
    super(props);

    this.form = createRef();

    const {
      mobile = '',
      nationality = defaultCountry,
      countries,
    } = this.props;

    const matchedDialingCountry = matchDialingCode(mobile, countries, nationality);

    this.state = {
      disableSelectDefaults: {
        title: false,
      },
      passportRequired: countries.find(matchCountry(nationality)).passportRequired,
      mobile,
      dialingCountryCode: matchedDialingCountry.countryCode,
    };

    if (mobile.startsWith((matchedDialingCountry.dialingCode))) {
      this.state.mobile = mobile.replace(matchedDialingCountry.dialingCode, '');
    }

    this.handleSubmit = this.handleSubmit.bind(this);
    this.validateTitleWrapper = this.validateTitleWrapper.bind(this);
    this.validateEmailWrapper = this.validateEmailWrapper.bind(this);
    this.isPassportRequired = this.isPassportRequired.bind(this);
    this.validateRegularGuestsForm = this.validateRegularGuestsForm.bind(this);
  }

  handleSubmit(values, actions) {
    const {
      setNotificationMessage,
      updateProfileStore,
      setEditView,
      updateProfile,
    } = this.props;
    const payload = this.createPayload(values);

    updateProfile(payload)
      .then(() => {
        setNotificationMessage(REGULAR_GUESTS_VIEW);
        actions.setSubmitting(false);
        updateProfileStore(payload);
        setEditView(DEFAULT_VIEW);
      }).catch(() => {
        setNotificationMessage(REGULAR_GUESTS_ERROR);
      });
  }

  createPayload({
    title,
    firstName,
    lastName,
    email,
    passportNumber,
    nationality,
    dialingCountryCode,
    mobile: mobileNumber,
  }) {
    const {
      profile,
      countries,
      index,
      email: existingEmail,
    } = this.props;

    const { additionalGuests = [] } = profile;

    const { dialingCode } = countries.find(country => (
      dialingCountryCode === country.countryCode));

    const mobile = `${dialingCode}${mobileNumber}`;

    const guestDetails = {
      title,
      firstName,
      lastName,
      email,
      passportNumber,
      nationality,
      mobile,
    };

    if (existingEmail) additionalGuests.splice(index, 1, guestDetails);
    else additionalGuests.push(guestDetails);

    return {
      ...profile,
      additionalGuests,
    };
  }

  isPassportRequired(e) {
    const { value } = e.target;
    const { countries } = this.props;

    this.setState({
      passportRequired: countries.find(matchCountry(value)).passportRequired,
    });
  }

  validateEmailWrapper(email) {
    const { profile: { additionalGuests = [] } } = this.props;
    const guestsEmails = additionalGuests.map(guest => guest.email);
    const emailChanged = this.form.current.initialValues.email !== email;
    return validateEmail(email, guestsEmails, emailChanged);
  }

  validateTitleWrapper(title) {
    const { disableSelectDefaults } = this.state;
    const result = validateTitle(title);

    this.setState({
      disableSelectDefaults: {
        ...disableSelectDefaults,
        title: !!title,
      },
    });

    return result;
  }

  validateRegularGuestsForm({
    title,
    firstName,
    lastName,
    email,
    passportNumber,
    mobile,
  }) {
    const errors = {};

    const { passportRequired } = this.state;

    errors.title = this.validateTitleWrapper(title);
    errors.firstName = validateFirstName(firstName);
    errors.lastName = validateLastName(lastName);
    errors.email = this.validateEmailWrapper(email);
    errors.mobile = validatePhoneNumber(mobile, 'profile.error.telephone');

    if (passportRequired) {
      errors.passportNumber = validatePassportNumber(passportNumber);
    }

    Object.keys(errors).forEach((error) => {
      if (errors[error] === undefined) delete errors[error];
    });

    return errors;
  }

  render() {
    const {
      app: {
        editViewName,
      },
      dictionary,
      countries,
      title = '',
      firstName = '',
      lastName = '',
      email = '',
      nationality = defaultCountry,
      passportNumber = '',
      index,
      setEditView,
    } = this.props;

    const {
      mobile,
      dialingCountryCode,
      passportRequired,
      disableSelectDefaults,
    } = this.state;


    return (
      <Fragment>
        <BaseStyle component="div">
          {editViewName === `REGULAR_GUESTS_VIEW_${index}` && (
            <Typography component="h4" className="pi-color-greyscale-darkest">
              {dictionary['guests.form.title.edit']}
            </Typography>
          )}
          { editViewName === 'ADD_GUEST_VIEW' && (
            <Typography component="h4" className="pi-color-greyscale-darkest">
              {dictionary['guests.form.title.add']}
            </Typography>
          )}
          <Notification priority={3} variant="info" ariaLive="off" show className="mb2">
            {dictionary['guests.notification']}
          </Notification>
        </BaseStyle>
        <Formik
          data-test="regular-guests-form-wrapper"
          onSubmit={this.handleSubmit}
          ref={this.form}
          initialValues={{
            title,
            firstName,
            lastName,
            email,
            nationality,
            passportNumber,
            dialingCountryCode,
            mobile,
          }}
          validate={this.validateRegularGuestsForm}
          render={formikProps => (
            <RegularGuestsForm
              {...formikProps}
              dictionary={dictionary}
              countries={countries}
              setEditView={setEditView}
              passportRequired={passportRequired}
              isPassportRequired={this.isPassportRequired}
              disableSelectDefaults={disableSelectDefaults}
            />
          )}
        />
      </Fragment>
    );
  }
}

RegularGuestsFormContainer.propTypes = {
  app: object.isRequired,
  profile: object.isRequired,
  dictionary: object.isRequired,
  countries: array.isRequired,
  updateProfileStore: func.isRequired,
  setNotificationMessage: func.isRequired,
  title: string,
  firstName: string,
  lastName: string,
  email: string,
  nationality: string,
  passportNumber: string,
  mobile: string,
  index: number,
  setEditView: func,
  updateProfile: func.isRequired,
};

export default RegularGuestsFormContainer;
