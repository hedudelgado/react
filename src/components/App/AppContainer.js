import { connect } from 'react-redux';
import { getSettings as getSettingsAction } from './appActions';
import { closeModal } from './modalActions';

import App from './App';

const mapStateToProps = ({
  modal: {
    display: displayModal,
    content: modalContent,
  },
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userProfile: {
    userDataReceived,
    profile,
  },
}) => ({
  profile,
  displayModal,
  modalContent,
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userDataReceived,
});

const mapDispatchToProps = dispatch => ({
  getSettings:
    (email, sessionId, business) => dispatch(getSettingsAction(email, sessionId, business)),
  handleModalClose: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
