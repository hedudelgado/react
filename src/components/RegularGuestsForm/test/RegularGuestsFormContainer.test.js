import React from 'react';
import { shallow } from 'enzyme';
import RegularGuestsFormContainer from '../RegularGuestsFormContainer';
import { matchDialingCode } from '../../../utils/matchDialingCode';
import { DEFAULT_VIEW, REGULAR_GUESTS_ERROR, REGULAR_GUESTS_VIEW } from '../../App/appActions';
import {
  validateTitle,
  validateFirstName,
  validateLastName,
  validateEmail,
  validatePassportNumber,
  validatePhoneNumber,
} from '../../../utils/validations';

jest.mock('../../../utils/validations', () => ({
  __esModule: true,
  validateTitle: jest.fn(),
  validateFirstName: jest.fn(),
  validateLastName: jest.fn(),
  validateEmail: jest.fn(),
  validatePassportNumber: jest.fn(),
  validatePhoneNumber: jest.fn(),
}));

describe('RegularGuestsFormContainer Component', () => {
  const DEFAULT_COUNTRY = {
    countryCode: 'GB',
    dialingCode: '+44',
  };

  let wrapper;
  let props;
  const updateProfile = jest.fn(() => Promise.resolve({}));

  beforeEach(() => {
    props = {
      updateProfile,
      app: {
        editViewName: 'test',
      },
      profile: {
        contactDetail: {
          title: 'Mr',
          firstName: 'Test',
          lastName: 'testing',
          email: 'test@whitbread.com',
          telephone: '11111111111',
          mobile: '+4900000000000',
          nationality: 'GB',
          passport: {
            number: '',
            countryOfIssue: '',
          },
          carRegistration: '',
          address: {
            line1: 'Random house',
            line2: 'Random street',
            line3: '',
            line4: 'LONDON',
            postCode: 'EC1N 3TD',
            countryCode: 'D',
            type: 'HOME',
            companyName: '',
          },
        },
        additionalGuests: [
          {
            title: 'Miss',
            firstName: 'Kentucky',
            lastName: 'Fried Chicken',
            email: 'kfc@test.com',
            telephone: '1111111111',
            mobile: '',
            nationality: 'GB',
            carRegistration: '',
          },
        ],
      },
      dictionary: {
        'guests.form.title.edit': 'Edit regular guest',
        'guests.form.title.add': 'Add regular guest',
        'guests.notification': 'Make sure you have permission to save this guestâ€™s details.',
        'guests.button.add': 'Add regular guest',
      },
      countries: [
        {
          countryCode: 'GB',
          countryCodeISO: 'GB',
          countryLegend: 'United Kingdom (the)',
          passportRequired: false,
          dialingCode: '+44',
          flagImg: '/content/dam/global/flags/United-Kingdom.png',
        },
        {
          countryCode: 'D',
          countryCodeISO: 'DE',
          countryLegend: 'Germany',
          passportRequired: true,
          dialingCode: '+49',
          flagImg: '/content/dam/global/flags/Germany.png',
        },
        {
          countryCode: 'GR',
          countryCodeISO: 'GR',
          countryLegend: 'Greece',
          passportRequired: true,
          dialingCode: '+30',
          flagImg: '/content/dam/global/flags/Greece.png',
        },
        {
          countryCode: 'TEST',
          countryCodeISO: 'TEST',
          countryLegend: 'Test',
          passportRequired: true,
          dialingCode: '+499',
          flagImg: 'test.png',
        },
      ],
      setNotificationMessage: jest.fn(),
      updateProfileStore: jest.fn(),
      setEditView: jest.fn(),
    };

    wrapper = shallow(<RegularGuestsFormContainer {...props} />);
    jest.clearAllMocks();
  });

  describe('Renders Regular Guests Form Container Component Elements', () => {
    it('should render the component', () => {
      expect(wrapper.find('Formik').dive()).toMatchSnapshot();
    });

    it('should render the notification', () => {
      expect(wrapper.find('Notification')).toHaveLength(1);
    });

    it('should render the edit guest title when on edit guest view', () => {
      const newProp = {
        app: {
          editViewName: 'REGULAR_GUESTS_VIEW_1',
        },
        index: 1,
      };
      wrapper = shallow(<RegularGuestsFormContainer {...props} {...newProp} />);
      expect(wrapper.find('Typography')).toHaveProp('children', 'Edit regular guest');
    });

    it('should render the add regular guest title when on add guest view', () => {
      const biggerProps = {
        app: {
          editViewName: 'ADD_GUEST_VIEW',
        },
        index: 1,
      };
      wrapper = shallow(<RegularGuestsFormContainer {...props} {...biggerProps} />);
      expect(wrapper.find('Typography')).toHaveProp('children', 'Add regular guest');
    });

    it('should set state false when a commonwealth country has been selected', () => {
      const event = { target: { value: 'GB' } };
      wrapper.instance().isPassportRequired(event);
      expect(wrapper.state().passportRequired).toBeFalsy();
    });
  });

  describe('Regular Guests Form Container initial values', () => {
    it('should get GB country code if there\'s no + indicator on mobile number', () => {
      expect(matchDialingCode('99999999', props.countries, 'MOON')).toEqual(DEFAULT_COUNTRY);
    });

    it('should match with address countryCode first', () => {
      expect(matchDialingCode('+4999999999', props.countries, 'D')).toBe(props.countries[1]);
    });

    it('should return strongest match if it doesn\'t match with address country code', () => {
      expect(matchDialingCode('+4999999999', props.countries, 'GB')).toBe(props.countries[3]);
    });

    it('should default to GB if there are no matches at all', () => {
      expect(matchDialingCode('+4123999999999', props.countries, 'GB')).toEqual(DEFAULT_COUNTRY);
    });

    it('should strip dialing code from mobile number', () => {
      const testingProps = {
        mobile: '+4400000000000',
      };

      wrapper = shallow(<RegularGuestsFormContainer {...testingProps} {...props} />);
      expect(wrapper.state().mobile).toBe('00000000000');
    });
  });

  describe('Regular Guests Form Container validation', () => {
    const current = {
      initialValues: {
        email: 'james.freeman@gmail.com',
      },
    };

    it('should validate fields with correct values when adding a new regular guest', () => {
      const validationProps = {
        ...props,
        profile: {
          contactDetail: {
            title: 'Mr',
            firstName: 'Test',
            lastName: 'testing',
            email: 'test@whitbread.com',
            telephone: '11111111111',
            mobile: '+4900000000000',
            nationality: 'GB',
            passport: {
              number: '',
              countryOfIssue: '',
            },
            carRegistration: '',
            address: {
              line1: 'Random house',
              line2: 'Random street',
              line3: '',
              line4: 'LONDON',
              postCode: 'EC1N 3TD',
              countryCode: 'D',
              type: 'HOME',
              companyName: '',
            },
          },
        },
        app: {
          editViewName: 'ADD_GUEST_VIEW',
        },
      };

      const values = {
        title: 'Mr',
        firstName: 'Morgan',
        lastName: 'Freeman',
        email: 'morgan.freeman@testmail.com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '1234567890',
      };

      wrapper = shallow(<RegularGuestsFormContainer {...validationProps} />);
      wrapper.instance().form.current = current;

      wrapper.setState({ passportRequired: true });

      wrapper
        .find('[data-test="regular-guests-form-wrapper"]')
        .props()
        .validate(values);

      expect(validateTitle).toHaveBeenCalledWith('Mr');
      expect(validateFirstName).toHaveBeenCalledWith('Morgan');
      expect(validateLastName).toHaveBeenCalledWith('Freeman');
      expect(validateEmail).toHaveBeenCalledWith('morgan.freeman@testmail.com', [], true);
      expect(validatePassportNumber).toHaveBeenLastCalledWith('1234567890');
      expect(validatePhoneNumber).toHaveBeenCalledWith('1234567890', 'profile.error.telephone');
    });

    it('should validate fields with correct values when adding another regular guest', () => {
      const validationProps = {
        ...props,
        app: {
          editViewName: 'ADD_GUEST_VIEW',
        },
      };

      const values = {
        title: 'Mr',
        firstName: 'Morgan',
        lastName: 'Freeman',
        email: 'morgan.freeman@testmail.com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '1234567890',
      };

      wrapper = shallow(<RegularGuestsFormContainer {...validationProps} />);
      wrapper.instance().form.current = current;
      wrapper.setState({ passportRequired: true });
      wrapper
        .find('[data-test="regular-guests-form-wrapper"]')
        .props()
        .validate(values);

      expect(validateTitle).toHaveBeenCalledWith('Mr');
      expect(validateFirstName).toHaveBeenCalledWith('Morgan');
      expect(validateLastName).toHaveBeenCalledWith('Freeman');
      expect(validateEmail).toHaveBeenCalledWith('morgan.freeman@testmail.com', ['kfc@test.com'], true);
      expect(validatePassportNumber).toHaveBeenLastCalledWith('1234567890');
      expect(validatePhoneNumber).toHaveBeenCalledWith('1234567890', 'profile.error.telephone');
    });

    it('should pass all validation and not display any error messages', () => {
      const values = {
        title: 'Mr',
        firstName: 'Morgan',
        lastName: 'Freeman',
        email: 'morgan.freeman@testmail.com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '1234567890',
      };

      wrapper.instance().form.current = current;

      const validationResult = wrapper
        .find('[data-test="regular-guests-form-wrapper"]')
        .props()
        .validate(values);

      expect(validationResult).toEqual({
        firstName: undefined,
        lastName: undefined,
        email: undefined,
        passportNumber: undefined,
        mobile: undefined,
      });
    });

    it('should return the error messages based on incorrect values', () => {
      const values = {
        title: 'Mr',
        firstName: 'Morgan!',
        lastName: 'Freeman1',
        email: 'morgan.freeman@com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '!1234567890',
      };

      validateTitle.mockImplementationOnce(() => 'Please select a title');
      validateFirstName.mockImplementationOnce(() => 'Please enter a valid first name');
      validateLastName.mockImplementationOnce(() => 'Please enter a valid last name');
      validateEmail.mockImplementationOnce(() => 'Please enter a valid email');
      validatePassportNumber.mockImplementationOnce(() => 'Please enter a valid passport number');
      validatePhoneNumber.mockImplementationOnce(() => 'Please enter a valid number');

      wrapper.instance().form.current = current;
      wrapper.setState({ passportRequired: true });

      const validationResult = wrapper
        .find('[data-test="regular-guests-form-wrapper"]')
        .props()
        .validate(values);

      expect(validationResult).toEqual({
        title: 'Please select a title',
        firstName: 'Please enter a valid first name',
        lastName: 'Please enter a valid last name',
        email: 'Please enter a valid email',
        passportNumber: 'Please enter a valid passport number',
        mobile: 'Please enter a valid number',
      });
    });
  });

  describe('Regular Guests Form Container submission', () => {
    it('should trigger handleSubmit function when RegularGuestsFormContainer is submitted', () => {
      const onSubmitMock = jest.fn();

      wrapper.instance().handleSubmit = onSubmitMock;
      wrapper.instance().forceUpdate();
      wrapper.find('Formik').simulate('submit');
      expect(onSubmitMock).toHaveBeenCalledTimes(1);
    });

    it('should execute Profile update success', async () => {
      const setSubmitting = jest.fn();
      const createPayload = jest.fn();

      const values = {
        title: 'Mr',
        firstName: 'Morgan',
        lastName: 'Freeman',
        email: 'morgan.freeman@testmail.com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '1234567890',
      };

      wrapper.instance().handleSubmit.apply({ props, createPayload }, [
        values,
        { setSubmitting },
      ]);

      await Promise.all([updateProfile]).then(() => {
        expect(props.setNotificationMessage).toHaveBeenCalledTimes(1);
        expect(props.updateProfileStore).toHaveBeenCalledTimes(1);
        expect(props.setEditView).toHaveBeenCalledTimes(1);
      });
    });

    it('should set ERROR notification when updateProfile fails ', async () => {
      updateProfile.mockImplementationOnce(jest.fn(() => Promise.reject({})));

      const values = {
        title: 'Mr',
        firstName: 'Morgan',
        lastName: 'Freeman',
        email: 'morgan.freeman@testmail.com',
        countryCode: 'TUR',
        passportNumber: '1234567890',
        dialingCountryCode: 'GB',
        mobile: '1234567890',
      };

      const createPayload = jest.fn();

      wrapper.instance().handleSubmit.apply({ props, createPayload }, [values, {}]);

      await Promise.all([updateProfile]).then(() => {
        expect(props.setNotificationMessage)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(REGULAR_GUESTS_ERROR);
        expect(props.updateProfileStore).not.toHaveBeenCalled();
      });
    });

    it('should generate correct payload if regular guests does not exist', async () => {
      const guestsProps = {
        ...props,
      };

      delete guestsProps.profile.additionalGuests;

      wrapper = shallow(<RegularGuestsFormContainer {...guestsProps} />);

      const values = {
        title: 'Mrs',
        firstName: 'Pizza',
        lastName: 'Hut',
        email: 'ph@test.com',
        mobile: '1111111111',
        dialingCountryCode: 'GB',
      };

      const actions = { setSubmitting: jest.fn() };
      wrapper.instance().handleSubmit(values, actions);

      await Promise.all([updateProfile]).then(() => {
        expect(props.setNotificationMessage)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(REGULAR_GUESTS_VIEW);
        expect(props.setEditView)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(DEFAULT_VIEW);
        expect(props.setEditView)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(DEFAULT_VIEW);
        expect(actions.setSubmitting)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(false);
        expect(props.updateProfileStore)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith({
            ...props.profile,
            additionalGuests: [
              {
                title: 'Mrs',
                firstName: 'Pizza',
                lastName: 'Hut',
                email: 'ph@test.com',
                mobile: '+441111111111',
              },
            ],
          });
      });
    });

    it('should generate correct payload when updating regular guest', async () => {
      const testProps = {
        email: 'kfc@test.com',
      };

      wrapper = shallow(<RegularGuestsFormContainer {...props} {...testProps} />);

      const values = {
        title: 'Miss',
        firstName: 'Kentucky',
        lastName: 'Extra Fried Chicken',
        email: 'kfc@test.com',
        telephone: '1111111111',
        mobile: '+441111111111',
        dialingCountryCode: 'GB',
      };

      const actions = { setSubmitting: jest.fn() };
      wrapper.instance().handleSubmit(values, actions);

      await Promise.all([updateProfile]).then(() => {
        expect(props.setNotificationMessage)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(REGULAR_GUESTS_VIEW);
        expect(props.setEditView)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(DEFAULT_VIEW);
        expect(props.setEditView)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(DEFAULT_VIEW);
        expect(actions.setSubmitting)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(false);
        expect(props.updateProfileStore)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith({
            ...props.profile,
            additionalGuests: [
              ...props.profile.additionalGuests,
            ],
          });
      });
    });
  });
});
