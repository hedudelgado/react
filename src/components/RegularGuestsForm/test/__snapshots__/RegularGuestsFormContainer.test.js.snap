// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegularGuestsFormContainer Component Renders Regular Guests Form Container Component Elements should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "dirty": false,
      "errors": Object {},
      "handleBlur": [Function],
      "handleChange": [Function],
      "handleReset": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "dialingCountryCode": "GB",
        "email": "",
        "firstName": "",
        "lastName": "",
        "mobile": "",
        "nationality": "GB",
        "passportNumber": "",
        "title": "",
      },
      "isSubmitting": false,
      "isValid": false,
      "isValidating": false,
      "registerField": [Function],
      "resetForm": [Function],
      "setError": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFormikState": [Function],
      "setStatus": [Function],
      "setSubmitting": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "status": undefined,
      "submitCount": 0,
      "submitForm": [Function],
      "touched": Object {},
      "unregisterField": [Function],
      "validate": [Function],
      "validateField": [Function],
      "validateForm": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validationSchema": undefined,
      "values": Object {
        "dialingCountryCode": "GB",
        "email": "",
        "firstName": "",
        "lastName": "",
        "mobile": "",
        "nationality": "GB",
        "passportNumber": "",
        "title": "",
      },
    }
  }
>
  <RegularGuestsForm
    countries={
      Array [
        Object {
          "countryCode": "GB",
          "countryCodeISO": "GB",
          "countryLegend": "United Kingdom (the)",
          "dialingCode": "+44",
          "flagImg": "/content/dam/global/flags/United-Kingdom.png",
          "passportRequired": false,
        },
        Object {
          "countryCode": "D",
          "countryCodeISO": "DE",
          "countryLegend": "Germany",
          "dialingCode": "+49",
          "flagImg": "/content/dam/global/flags/Germany.png",
          "passportRequired": true,
        },
        Object {
          "countryCode": "GR",
          "countryCodeISO": "GR",
          "countryLegend": "Greece",
          "dialingCode": "+30",
          "flagImg": "/content/dam/global/flags/Greece.png",
          "passportRequired": true,
        },
        Object {
          "countryCode": "TEST",
          "countryCodeISO": "TEST",
          "countryLegend": "Test",
          "dialingCode": "+499",
          "flagImg": "test.png",
          "passportRequired": true,
        },
      ]
    }
    dictionary={
      Object {
        "guests.button.add": "Add regular guest",
        "guests.form.title.add": "Add regular guest",
        "guests.form.title.edit": "Edit regular guest",
        "guests.notification": "Make sure you have permission to save this guestâ€™s details.",
      }
    }
    dirty={false}
    disableSelectDefaults={
      Object {
        "title": false,
      }
    }
    errors={Object {}}
    handleBlur={[Function]}
    handleChange={[Function]}
    handleReset={[Function]}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "dialingCountryCode": "GB",
        "email": "",
        "firstName": "",
        "lastName": "",
        "mobile": "",
        "nationality": "GB",
        "passportNumber": "",
        "title": "",
      }
    }
    isPassportRequired={[Function]}
    isSubmitting={false}
    isValid={false}
    isValidating={false}
    passportRequired={false}
    registerField={[Function]}
    resetForm={[Function]}
    setEditView={[MockFunction]}
    setError={[Function]}
    setErrors={[Function]}
    setFieldError={[Function]}
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    setFormikState={[Function]}
    setStatus={[Function]}
    setSubmitting={[Function]}
    setTouched={[Function]}
    setValues={[Function]}
    submitCount={0}
    submitForm={[Function]}
    touched={Object {}}
    unregisterField={[Function]}
    validateField={[Function]}
    validateForm={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
    values={
      Object {
        "dialingCountryCode": "GB",
        "email": "",
        "firstName": "",
        "lastName": "",
        "mobile": "",
        "nationality": "GB",
        "passportNumber": "",
        "title": "",
      }
    }
  />
</ContextProvider>
`;
